hipcheck/src/plugin/manager.rs

```diff
@@ -0,0 +1,121 @@
+use crate::hipcheck::plugin_client::PluginClient;
+use crate::plugin::{HcPluginClient, Plugin, PluginContext};
+use crate::{hc_error, Result, F64};
+use rand::Rng;
+use std::collections::HashSet;
+use std::ops::Range;
+use std::process::Command;
+use tokio::time::{sleep_until, Duration, Instant};
+
+pub struct PluginExecutor {
+	max_spawn_attempts: usize,
+	max_conn_attempts: usize,
+	port_range: Range<u16>,
+	backoff_interval: Duration,
+	jitter_pct: u8,
+	est_ports: HashSet<u16>,
+}
+impl PluginExecutor {
+	pub fn new(
+		max_spawn_attempts: usize,
+		max_conn_attempts: usize,
+		port_range: Range<u16>,
+		backoff_interval_micros: u64,
+		jitter_pct: u8,
+	) -> Result<Self> {
+		if jitter_pct > 100 {
+			return Err(hc_error!("jitter_pct must be <= 100, got {}", jitter_pct));
+		}
+		let backoff_interval = Duration::from_micros(backoff_interval_micros);
+		Ok(PluginExecutor {
+			max_spawn_attempts,
+			max_conn_attempts,
+			port_range,
+			backoff_interval,
+			jitter_pct,
+			est_ports: HashSet::new(),
+		})
+	}
+	fn get_available_port(&mut self) -> Result<u16> {
+		for i in self.port_range.start..self.port_range.end {
+			if !self.est_ports.contains(&i)
+				&& std::net::TcpListener::bind(format!("127.0.0.1:{i}")).is_ok()
+			{
+				return Ok(i);
+			}
+		}
+		Err(hc_error!("Failed to find available port"))
+	}
+	pub async fn start_plugin(&mut self, plugin: &Plugin) -> Result<PluginContext> {
+		let mut rng = rand::thread_rng();
+		// Plugin startup design has inherent TOCTOU flaws since we tell the plugin
+		// which port we expect it to bind to. We can try to ensure the port we pass
+		// on the cmdline is not already in use, but it is still possible for that
+		// port to become unavailable between our check and the plugin's bind attempt.
+		// Hence the need for subsequent attempts if we get unlucky
+		let mut spawn_attempts: usize = 0;
+		while spawn_attempts < self.max_spawn_attempts {
+			// Find free port for process. Don't use spawn_attempts, since
```

Comment:
This comment seems incomplete or it was cut off in editing -- was there more to say about why not to use `spawn_attempts` here?

Reply:
Fixed.